nsqlookupd: synchronize goroutines to avoid t.Log data races

benjsto opened the pull request:
Fixes the data race seen here: https://github.com/nsqio/nsq/issues/1157 by providing a mechanism for all handler goroutines in `nsqlookupd` to exit before tests finish.

I'm obviously new here so let me know if there are better/simpler ways to accomplish this.

ploxiln commented on nsqlookupd/lookup_protocol_v1.go:
```diff
@@ -27,9 +28,30 @@ func (p *LookupProtocolV1) IOLoop(conn net.Conn) error {
+		go func() {
+			line, e := reader.ReadString('\n')
+			readChan <- line
+			errChan <- e
```
I think it would be better to instead find a way to call `conn.Close()` when `exitChan` indicates. Here's one idea:

```go
doneChan := make(chan interface{})
defer close(doneChan)
go func() {
        select {
        case <-p.exitChan:
                conn.Close()
        case <-doneChan:
                break
        }
}
for {
        line, err = reader.ReadString('\n')
        ...
```

Here's a different way that does not involve an extra goroutine and channel for each connection:

```diff
--- a/nsqlookupd/nsqlookupd.go
+++ b/nsqlookupd/nsqlookupd.go
@@ -66,6 +66,7 @@ func (l *NSQLookupd) Main() error {
        tcpServer := &tcpServer{ctx: ctx}
        l.waitGroup.Wrap(func() {
                exitFunc(protocol.TCPServer(l.tcpListener, tcpServer, l.logf))
+               tcpServer.CloseAll()
        })
        httpServer := newHTTPServer(ctx)
        l.waitGroup.Wrap(func() {

--- a/nsqlookupd/tcp.go
+++ b/nsqlookupd/tcp.go
@@ -3,12 +3,14 @@ package nsqlookupd
 import (
        "io"
        "net"
+       "sync"
 
        "github.com/nsqio/nsq/internal/protocol"
 )
 
 type tcpServer struct {
        ctx *Context
+       conns sync.Map
 }
 
 func (p *tcpServer) Handle(clientConn net.Conn) {
@@ -41,9 +43,20 @@ func (p *tcpServer) Handle(clientConn net.Conn) {
                return
        }
 
+       p.conns.Store(clientConn.RemoteAddr(), clientConn)
+
        err = prot.IOLoop(clientConn)
        if err != nil {
                p.ctx.nsqlookupd.logf(LOG_ERROR, "client(%s) - %s", clientConn.RemoteAddr(), err)
-               return
        }
+
+       p.conns.Delete(clientConn.RemoteAddr())
+}
+
+func (p *tcpServer) CloseAll() {
+       p.conns.Range(func (k, v interface{}) bool {
+               v.(net.Conn).Close()
+               p.conns.Delete(k)
+               return true
+       })
 }
```

but I'm not sure if the overhead of the `sync.Map()` is something to worry about. probably not :)

ploxiln replied:
hmm, I think it would need a different way to call `tcpServer.CloseAll()` for my second proposal, ordering after `protocol.TcpServer()` returns seems like it shouldn't work

ploxiln replied:
a better way to call `CloseAll()` for that idea:

```diff
--- a/nsqlookupd/nsqlookupd.go
+++ b/nsqlookupd/nsqlookupd.go
@@ -18,6 +18,7 @@ type NSQLookupd struct {
        opts         *Options
        tcpListener  net.Listener
        httpListener net.Listener
+       tcpServer    *tcpServer
        waitGroup    util.WaitGroupWrapper
        DB           *RegistrationDB
 }
@@ -63,9 +64,9 @@ func (l *NSQLookupd) Main() error {
                })
        }
 
-       tcpServer := &tcpServer{ctx: ctx}
+       l.tcpServer = &tcpServer{ctx: ctx}
        l.waitGroup.Wrap(func() {
-               exitFunc(protocol.TCPServer(l.tcpListener, tcpServer, l.logf))
+               exitFunc(protocol.TCPServer(l.tcpListener, l.tcpServer, l.logf))
        })
        httpServer := newHTTPServer(ctx)
@@ -87,6 +88,7 @@ func (l *NSQLookupd) RealHTTPAddr() *net.TCPAddr {
 func (l *NSQLookupd) Exit() {
        if l.tcpListener != nil {
                l.tcpListener.Close()
+               l.tcpServer.CloseAll()
        }
```

benjsto replied:
Trying that now, I like it. Thank you for taking a look at this!

ploxiln replied:
thanks for the updates! I'm going to take a day or two to test it a bit myself

ploxiln commented on internal/protocol/tcp_server.go:
```diff
@@ -30,9 +33,18 @@ func TCPServer(listener net.Listener, handler TCPHandler, logf lg.AppLogFunc) er
+		wg.Add(1)
+
+		go func() {
+			defer wg.Done()
```
minor nit: you do not need `defer` because there are not multiple return paths from this function. So it's a tiny bit simpler and a tiny bit more efficient to just put this statement second, without `defer`.

ploxiln commented on nsqlookupd/tcp.go:
```diff
@@ -41,9 +43,20 @@ func (p *tcpServer) Handle(clientConn net.Conn) {
+func (p *tcpServer) CloseAll() {
+	p.conns.Range(func(k, v interface{}) bool {
+		v.(net.Conn).Close()
+		p.conns.Delete(k)
```
I think this `Delete()` is not really needed here, it'll be done at the end of `Handle()` anyway.

ploxiln commented:
worth noting: nsqd also uses `TCPServer`, and already closes all connections via `Topic.Close()` -> `Channel.Close()` -> `Consumer.Close()`

I think this is good. Please squash commits together before merge.

benjsto commented:
I assumed something like that was going on for nsqd's use of TCPServer but hadn't dug into it. Thanks for clarifying. Thanks for your help on this - I've removed that unnecessary Delete and squashed and rebased.
